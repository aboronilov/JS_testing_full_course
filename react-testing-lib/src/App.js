import logo from "./logo.svg";
import "./App.css";
import { useEffect, useState } from "react";

function App() {
  const [data, setData] = useState(null)
  const [toggle, setToggle] = useState(false)
  const [value, setValue] = useState('')

  const onClick = () => setToggle(value => !value)

  useEffect(()=>{
    setTimeout(()=>{
      setData({})
    }, 100)
  },[])

  return (
    <div>
      <h1 data-testid="value-elem">{value}</h1>
      {toggle === true && <div data-testid="toggle-elem">toggle</div>}
      {data && <div style={{color:'red'}}>data</div>}
      <h1>Hello world</h1>
      <button data-testid="toggle-btn" onClick={onClick}>Click me</button>
      <input type="text" placeholder="input value" onChange={event=>setValue(event.target.value)}></input>
    </div>
  );
}

export default App;
