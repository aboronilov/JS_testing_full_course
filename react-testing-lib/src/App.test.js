import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';

describe('TEST APP', ()=>{
  test('renders APP elements', () => {
    render(<App />);
    const h1Element = screen.getByText(/hello world/i);
    const btn = screen.getByRole('button')
    const input = screen.getByPlaceholderText(/input value/i)
    expect(h1Element).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
    expect(input).toMatchSnapshot();
  
    // можно пользоваться screen.debug()
  });

  test('does not render none element', () => {
    render(<App />);
    // find. Возвращает object завернутый в promise. Для асинхронных функций
    // get - если элемент не найден, тест падает с ошибкой. Возвращает object
    // query - для проверки отсутствия. Если не найден, вернет null - можно по нему проверить 
    const el = screen.queryByText(/fake text/i);
    expect(el).toBeNull()
  });

  test('renders async element', async () => {
    render(<App />); 
    screen.debug()
    const data = await screen.findByText(/data/i);
    expect(data).toBeInTheDocument()
    expect(data).toHaveStyle({color:'red'})
    screen.debug()
  });

  test('click event handling', async () => {
    render(<App />); 
    const btn = screen.getByTestId("toggle-btn")
    expect(screen.queryByTestId("toggle-elem")).toBeNull()

    // fireEvent - для работы с событиями

    fireEvent.click(btn)
    expect(screen.queryByTestId("toggle-elem")).toBeInTheDocument()
    fireEvent.click(btn)
    expect(screen.queryByTestId("toggle-elem")).toBeNull()
  });

  test('input event', async () => {
    render(<App />); 
    const input = screen.getByPlaceholderText(/input value/i)
    expect(screen.queryByTestId("value-elem")).toContainHTML('')
    fireEvent.input(input, {
      target: {value: 'test'}
    })
    expect(screen.queryByTestId("value-elem")).toContainHTML('test')
    expect(screen.queryByTestId("value-elem")).not.toContainHTML('fake')

    // userEvent - для имитации поведения пользователя (работа с клавиатурой)

    userEvent.type(input, 'test')
    expect(screen.queryByTestId("value-elem")).toContainHTML('test')
  });
})


