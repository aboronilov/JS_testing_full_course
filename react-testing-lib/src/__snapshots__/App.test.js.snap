// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST APP renders APP elements 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`renders APP 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;
